File: compositepattern/filesystem.py
from abc import ABC, abstractmethod
class FileSystem(ABC):
    @abstractmethod
    def ls(self):
        pass

class File(FileSystem):
    def __init__(self, name):
        self.fileName = name

    def ls(self):
        print("file name ", self.fileName)

class Directory(FileSystem):
    def __init__(self, name):
        self.directoryName = name
        self.fileSystemList = []

    def add(self, fileSystemObj):
        self.fileSystemList.append(fileSystemObj)

    def ls(self):
        print("Directory name ", self.directoryName)

        for fileSystemObj in self.fileSystemList:
            fileSystemObj.ls()

if __name__ == '__main__':
    movieDirectory = Directory("Movie")
    border = File("Border")
    movieDirectory.add(border)

    comedyMovieDirectory = Directory("ComedyMovie")
    hulchul = File("Hulchul")
    comedyMovieDirectory.add(hulchul)
    movieDirectory.add(comedyMovieDirectory)

    movieDirectory.ls()

File: compositepattern/filesystem.js
class FileSystem {
    ls() {
        throw new Error('Abstract method "ls()" must be implemented');
    }
}

class File extends FileSystem {
    constructor(fileName) {
        super()
        this.fileName = fileName
    }

    ls() {
        console.log('file name:', this.fileName)
    }
}

class Directory extends FileSystem {
    constructor(directoryName) {
        super()
        this.directoryName = directoryName
        this.fileSystemList = []
    }

    add(fileSystemObj) {
        this.fileSystemList.push(fileSystemObj)
    }

    ls() {
        console.log('Directory name:', this.directoryName)

        for (let fileSystemObj of this.fileSystemList) {
            fileSystemObj.ls()
        }
    }
}


const movieDirectory = new Directory("Movie")
const border = new File("Border")
movieDirectory.add(border)

const comedyMovieDirectory = new Directory("ComedyMovie")
const hulchul = new File("Hulchul")
comedyMovieDirectory.add(hulchul)
movieDirectory.add(comedyMovieDirectory)

movieDirectory.ls()

